### PR TITLE
Add textual information for Record and Union diff

### LINKED ISSUE #1846: Papercut: incorrect kind errors are somewhat cryptic
Every now and then I see errors along the lines of:

```
- < … : … >
+ { … : … }
```

This is a very symbol-dense, and not that obvious what it means (especially when it appears deep in a structure). And if you're still learning, you might not have written enough unions / record types to even recognise the meaning of the different braces.

For cases where the entire kind (not sure if that's the right word) is wrong, would it be possible to either replace this text or append a descriptive label to the kind? e.g.:

```
- < … : … > (a union)
+ { … : … } (a record)
```

### PR DESCRIPTION
**Pull request by TristanCacqueray**
This change improves error messages when argument types are wrong.

Fixes #1846

### CODE REVIEW

File: dhall/src/Dhall/Diff.hs [hunk 1]
```diff
@@ -556,6 +556,7 @@ skeleton (Record {}) =
     <>  ignore
     <>  " "
     <>  rbrace
+    <>  " (a record)"
```
Comment (Gabriel439) [inline dhall/src/Dhall/Diff.hs:559]:
```suggestion
    <>  " (a record type)"
```

Also, could you add a similar explanation for record literals, too?

**Comment by TristanCacqueray** [comment]
Example:
```
$ dhall <<< '<a | b>.a : { a : Bool }'

Use "dhall --explain" for detailed errors

Error: Expression doesn't match annotation

- { … : … } (a record)
+ < … : … > (an union)
```

File: dhall/src/Dhall/Diff.hs [hunk 2]
```diff
@@ -576,6 +577,7 @@ skeleton (Union {}) =
     <>  ignore
     <>  " "
     <>  rangle
+    <> " (an union)"
```
Comment (Gabriel439) [inline dhall/src/Dhall/Diff.hs:580]:
```suggestion
    <> " (a union type)"
```

**Comment by TristanCacqueray** [comment]
New examples:
```
$ dhall <<< '<a | b>.a : { a : Bool }'
Error: Expression doesn't match annotation

- { … : … } (a record type)
+ < … : … > (a union type)

$ dhall <<< '<a | b>.a : { a = Bool }'
Error: Expression doesn't match annotation

- { … = … } (a record)
+ < … : … > (a union type)

$ dhall <<< '{ a : Bool } : { a = Bool }'
Error: Expression doesn't match annotation

- { … = … } (a record)
+ Type
```

File: dhall/src/Dhall/Diff.hs (hunk 1 above)
Comment (TristanCacqueray) [reply dhall/src/Dhall/Diff.hs]: Oops indeed, thanks you!

Comment (Gabriel439) [reply dhall/src/Dhall/Diff.hs]: You're welcome! 🙂

File: dhall/src/Dhall/Diff.hs [hunk 3]
```diff
@@ -576,6 +578,7 @@ skeleton (Union {}) =
     <>  ignore
     <>  " "
     <>  rangle
+    <> " (a union type)"
```
Comment (timbertson) [inline dhall/src/Dhall/Diff.hs:581]: Should this be "a union"? It's talking about an instance of a union type, I think?

**Comment by timbertson** [comment]
Thanks! Don't suppose you could add "(a function)" to the `Lam` skeleton too? I think that's rendered as `… → …` which is a bit less cryptic but could still benefit from some words imo.

File: dhall/src/Dhall/Diff.hs (hunk 3 above)
Comment (Gabriel439) [reply dhall/src/Dhall/Diff.hs:581]: @timbertson: No, this is a union type.  A union value is of the form `UnionType.Alternative x`

**Comment by TristanCacqueray** [comment]
@timbertson last commit adds "(a function)" message to `Lam`, though that doesn't work well:

```
$ dhall <<< '(\(a : Bool) -> a) : Bool'
Error: Expression doesn't match annotation

- Bool
+ … → …

$ dhall <<< 'Bool : (\(a : Bool) -> a)'
Error: Expression doesn't match annotation

- λ(… : …) → … (a function)
+ Type
```

And that wouldn't work for return type:

```
$ dhall <<< '(\(a : Bool) -> 42) : forall (a : Bool) -> Bool'
Error: Expression doesn't match annotation

  …
→ - Bool
  + Natural
```

I guess "(a function)" needs to be added to `Pi` instead, but I'm not entirely sure why it shouldn't be "(a function type)" :)

**Comment by sjakobi** [comment]
`(a function)` seems fine to me for `Lam`. And for consistency, `Pi` should get `(a function type)`.

Thanks BTW – I think this will make some errors much easier to understand! :)

**Comment by sjakobi** [comment]
BTW, Hydra OOM'd again (error 137), with GHC 8.4.3. This has been happening repeatedly in the last few days, possibly due to more contributions during ZuriHac.

I'm wondering though whether it might be time to upgrade to a machine with more RAM. If costs are an issue, maybe it could be financed from the donations?

**Comment by Gabriel439** [comment]
@sjakobi: Cost would be an issue 🙂.  I currently pay $40 / month for a Linode with 8 GB of RAM.  The next tier with 16 GB would be $80 / month, which would consume a significant fraction of our donation budget.

I would personally wait to see if it's still a problem now that Zurihac's over

**Comment by sjakobi** [comment]
Sure. Of course we could also look at cheaper providers, as for example [`static-haskell-nix` did](https://opencollective.com/static-haskell-nix/updates/build-server-funding-goal-reached), but it's all work that isn't strictly necessary when the existing server can mostly shoulder the load.

**Comment by sjakobi** [comment]
@Gabriel439 I believe the OOM'ing Hydra job needs to be restarted: https://hydra.dhall-lang.org/build/61554/nixlog/2

**Pull request merged**